### ISSUE TITLE
Do yaw optimizations with one rotor point and refine with more points to increase computational speed

### ISSUE THREAD

**Issue by rafmudaf** [NREL/floris#289]

**Comment by Bartdoekemeijer** [comment]
When evaluating example case ``08_opt_yaw_single_ws.py`` with
```
solver:
  type: turbine_grid
  turbine_grid_points: 1
```
and
```
enable_transverse_velocities: true
```
I find an error in ``wake_deflection/gauss.py:calculate_transverse_velocity()``, line 411, having to do with the variable ``dudz_initial`` and function ``np.gradient``. I think we need at least two vertical rotor coordinates to calculate the vertical velocity induced by yaw. Note that ``enable_transverse_velocities`` significantly impacts the flow and optimal yaw angle solutions. This means we always need at least ``turbine_grid_points: 2``.

Timing the yaw optimization for a 25-turbine farm for wind directions 0 to 360 deg:
```
turbine_grid_points: 3. Time spent: 20.6 s [25 turbines]
turbine_grid_points: 2. Time spent: 17.5 s [25 turbines]
```

We see about a 15% reduction in computation time. I don't think a 15% reduction is worth it compared to the increased code complexity and potential loss in modeling accuracy we may have by going down to ``turbine_grid_points: 2``.

Hence, I'd say let us not further pursue this. What are your thoughts on this @paulf81 @rafmudaf ?

**Comment by rafmudaf** [comment]
Yes, I saw this as well. We could use a single value there if one point in the z-direction, but I'm not sure what it should be.

**Comment by paulf81** [comment]
I think this matches my expectation and discussions we had at chez @bayc .  One point is probably acceptable for non-wake steering results (or those based on Jimenez?) but all the -curl-hybrid models need a gradient and that brought us to defaulting to 3.  I was thinking of making a separate example showing a faster AEP-only calculation with one point maybe to make this point more clearly?

**Comment by Bartdoekemeijer** [comment]
@paulf81 interesting. So perhaps a better use case would be layout optimization (without wake steering) using a single rotor point? For simple AEP calculations perhaps it doesn't matter too much, since it's already so fast.

**Comment by paulf81** [comment]
Layout optimization, or maybe more generally, any process that functions like, change some value, recompute AEP, repeat (iterating turbine height maybe?) will all benefit from faster options.  Just as a check I made a short AEP calculation comparison using the new CC model for a 50 turbine farm.  Results:

 - turbine_grid_points: 3
 - AEP: 854.03 GWh
 - time: 76.6s

- turbine_grid_points: 2
- AEP: 853.95 GWh
- time: 28.4 s

  That could represent an important speed boost to certain operations, and still room to improve with a drop to 1 point, I think we'd just need to only throw the gradient error if turbines yawed?

**Comment by Bartdoekemeijer** [comment]
That is insightful. So there are significant performance improvements with fewer grid points. I'm thinking, perhaps this is an advanced use case. I don't think this functionality necessarily needs to be included in the optimizers. Rather, it would be one strategy of how the optimizers are used. For example,

1. One could get close to an optimal solution with `turbine_grid_points: 1` or `turbine_grid_points: 2`, and then
2. Re-do the optimization with `turbine_grid_points: 3` and the initial condition `x0` being the optimal solution from step 1.

Perhaps we can simply demonstrate such a use case in an example file? It seems to me that we wouldn't really need to modify any of the optimization classes.

**Comment by misi9170** [comment]
I think we can treat this as stale, with various improvements in overall speed/performance in subsequent FLORIS releases. If speed of yaw optimizations is still a pressing issue, we could consider reopening.
